iothub/device/samples/getting started/MessageReceiveSample/MessageReceiveSample.cs

```diff
@@ -104,13 +110,15 @@ private async Task ReceiveC2dMessagesPollingAndCompleteAsync(CancellationToken c
 
         private async Task OnC2dMessageReceivedAsync(Message receivedMessage, object _)
         {
+            await _processMessageSemaphore.WaitAsync().ConfigureAwait(false);
```

Comment:
I'd recommend passing in a cancellation token to this `WaitAsync()` so that we aren't waiting indefinitely.

Semaphores and locks can cause deadlocks if not implemented correctly, i.e. we need to be careful how we expect them to be implemented in both the happy path and also corner cases where unexcepted exceptions are encountered. That does not mean that we should never use them, but it is definitely worth investigating alternate approaches. 

Are there any alternative approaches to solve this issue?

Reply:
Discussed it with David in the morning and he suggested using semaphores as its a smaller change compared to having concurrent queue.